enhancements/coreos-bootimages.md

```diff
@@ -0,0 +1,143 @@
+
+### Stream metadata format
+
+As part of unifying Fedora CoreOS and RHEL CoreOS, we have standardized on the "stream metadata" format used by FCOS.  More in [FCOS docs](https://docs.fedoraproject.org/en-US/fedora-coreos/getting-started/) and [this RHCOS issue](https://github.com/openshift/os/issues/477).
+
+There is a new [stream-metadata-go](https://github.com/coreos/stream-metadata-go) library to consume this data.
+
+### Adding a new git repository and release image component + configmap with this data
+
+A new git repository https://github.com/openshift/rhel-coreos-bootimages will be created (based on [an existing prototype repository](https://github.com/cgwalters/rhel-coreos-bootimages)).  It will have the stream JSON.
+
+Additionally, this repository will be included in the release image as `rhel-coreos-bootimages` and it will use the CVO to install a `configmap/coreos-bootimages` in the `openshift-machine-config-operator` namespace.
+
+### Update `oc adm release new` to patch the installer binary to inject this data
+
+The installer needs CoreOS to start at all; today the bootimages are used for the bootstrap node
+and control plane which are created by the installer directly.  In order to avoid duplicating the
+metadata, change `oc adm release new` to *binary patch* the included `openshift-install`
+binary, in the same way we [do for the release image](https://github.com/openshift/installer/pull/1422).
+
+The RHCOS team will take over maintenance, CI, and automation of `rhel-coreos-bootimages`.  For OKD,
+the OKD team will adapt their existing tooling to create something similar for Fedora CoreOS.
+
+### Add `oc adm release info --coreos-bootimages quay.io/openshift-release-dev/ocp-release:4.7.0-x86_64`
+
+This command will extract the CoreOS bootimage stream metadata from a release image and should be used by UPI installs (including our many existing automated tests for UPI).
+
+Additionally, we will add `oc adm release coreos-download -p openstack` which will e.g. download the OpenStack `.qcow2`, including verifying its integrity.
+
+### Data available at https://mirror.openshift.com
+
+The way we deliver bootimages at http://mirror.openshift.com/pub/openshift-v4/x86_64/dependencies/rhcos/4.7/latest/ is not friendly to automation.  By placing the stream metadata JSON there, we gain a standardized machine-readable format.
+
+The ART team will synchronize the stream metadata with the data embedded in the latest release image for a particular OpenShift minor.
+
+
+## Motivation
+
+### Goal 1: Including metadata in the cluster so that we can later write automated IPI/machineAPI updates
+
+Lay the groundwork so that one or both of the MCO or machineAPI operators can act on this data, and e.g. in an AWS cluster update the machineset to use a new base AMI.
```

Comment:
want to understand what sort of changes MCO would need to do? bootimages today are ideally handled by machine-api, right?

Reply:
This topic bridges the two I think.  Indeed the machineAPI naturally would be a good place to update machinesets.  But on the flip side, this stuff is completely CoreOS specific (like a lot of the MCO), whereas the machineAPI is a lot more generic (because it's forked from the upstream generic Kube cluster API).

Also if the goal is to use this to skip the firstboot os upgrade, we would want some careful coordination around analyzing he diff between machine-os-content and the bootimages to ensure we don't actually need to do the in-place update.  IOW there's no point to doing the machineset update if we still need to firstboot reboot - so that's MCO knowledge.

Or as I commented below, perhaps this is even a new (small) operator.  I don't have a really strong opinion to be clear.  This enhancement just calls for injecting the data so we can later do that work.

(But, the configmap is placed in the `openshift-machine-config-operator` namespace; if you guys object to that or want it changed please say so!)

Reply:
> Also if the goal is to use this to skip the firstboot os upgrade, we would want some careful coordination around analyzing he diff between machine-os-content and the bootimages to ensure we don't actually need to do the in-place update. IOW there's no point to doing the machineset update if we still need to firstboot reboot - so that's MCO knowledge.

agree with making MCO smarter here to look at OSTree content rather than just checking the OSImageURL.
> 
> This enhancement just calls for injecting the data so we can later do that work.

right

> (But, the configmap is placed in the `openshift-machine-config-operator` namespace; if you guys object to that or want it changed please say so!)

either machineAPI or MCO namespace sounds reasonable place for this. Based on your reasoning about MCO over machineAPI, I don't see any issue here.